Add support for informers

macsux opened the pull request:
This PR adds support for informers that form building block for creating controllers. It follows the spirit of original informers, but differs in implementation found in Go and Java in a few significant ways.

- The entire library is built on reactive paradigm using Reactive Extensions. RX is a natural fit as we're dealing with streams of events, allowing leverage powerful built in functional folds over live streams of object change data. This offers advantages in:

  - intuitive LINQ like operators to filter / transform and collect live information being streamed when attached to k8s
  - reuse of single connection for internal multicasting
  - scheduler abstractions to provide multithreaded synchronization and concurrency
  - virtual time for unit testing

- Flexible abstraction model based on IInformer interface that allows modeling observable resources over resources other then K8S. This allows observing a SET of resources of similar type while conforming to same logical interface, easily allowing connecting one resource type to another. IInformer abstracts out a List/Watch operations into a single observable stream of events. Each event in the stream includes a bitvise enum mask that describes event purpose and how receivers of event should treat it, as well the current state of the object. 

  - List operations are marked by event type "Reset", with first and last elements having ResetStart/ResetEnd flags. Reset also implies that any previously tracked cache state should be discard and new one used.
  - If list operation is empty and is followed by watch, ResetEmpty flag is used to denote transition point
  - Add/Modify/Delete flags denote changes to existing state
  - (Re)Sync events to denote periodic refreshes of same data already sent

  The intended vision is to allow reuseable set of informers for other APIs to be created and used as building blocks for creating controllers. For example a reusable informer can be written for twitter feed, leveraging push based websocket features. Legacy APIs without efficient change detection such an equivalent of Watch can be abstracted to provide push based events via polling. 

- Concrete implementations of IInformer follow individual API rules, as well as provide reasonable failure recovery models.

- Generic List operation for `Kubernates` object that can work with any existing object or CRD.

  - Future enhancement will add similar features for modifying the state

- Included K8S informer provides supports

  - List followed by Watch
  - Fault tolerance policy with default backoff retry
  - Seamless connection recovery over transient error codes, including support for bookmarks
  - Server side support for namespace filtering
  - Future todo: server side label filtering

- Shared informer that is not K8S specific acts as an in-memory broadcaster and cache synchronizer. It follows lazy loading semantics, attaching multiple logical in-memory subscribers connection to single master Informer. This allows multiple controllers to share a single physical connection to the API server. Connection is automatically established when first subscription is made, and detached when it's dropped. List type queries for internal subscribers are simulated from local synchronized cache

- Shared informer factory that dynamically allocates specific shared informer based on resource type and server side query

- Support for shared versioned cache that allows resources that follow versioning semantics to be reused across multiple logical caches. This is useful when two informers are monitoring a subset of same resource where overlaps are possible (such as a server side label query), but one does not want to maintain two sets of objects with same key/version on the heap.

- Logical reactive fold helper that allows new caches to be created based on arbitrary key selector. This is an alternative to indexed caches that are used in Go implementation. This is a preferred model because it doesn't require complex global cache synchronization issues that exist with Go implementation.

- A functional fold for detection of discrepancies between relists and simulation of missed events. For example a List operation that returned [ObjectA.v1, ObjectB.v1] followed by disconnect and a relist that yields only [ObjectA.v1], it computes that the discrepency with local cache on ObjectB is a missed delete operation and changes the stream accordingly. This can be leveraged as needed simply by adding `.ComputeMissedEventsBetweenResets` extension method to observable.

- A worker with "events per object key" semantics, allowing processing of all events related to single event as a single unit of work. This offers similar functions as Go's DeltaFifo implementation. This is based on TPL Dataflows which are included in .NET Core and offer powerful in-memory actor model which can be used to develop additional queuing and processing semantics.

- Extensive unit test coverage 

  - Virtual time based on RX test scheduler
  - Reusable scheduled event "scenarios" for writing theories
  - Improved test readability via use of NSubstitute and FluentAssertions.
  - Integration tests with mocking of real K8S via WireMock.NET. This is more powerful then existing mock object which doesn't currently support mocking multiple requests or streaming results.

- Sample informer project that demonstrates how to use informer as a starting point to write a simple controller

Things to note

- this PR bumps up required .NET Framework version to 4.6.1 as some of the libraries target NetStandard 2.0.
- The current .NET Kubernetes library is becoming quite monolithic and has too many dependencies. I propose that it be modularized into separate nuget packages. I'm happy to refactor informers into its own package if this is what maintainers want to do. The next phase for controllers will introduce the need for even more dependencies, so it makes sense to do this now. See sample project on where this could be taken

macsux commented:
/assign @krabhishek8260

macsux commented:
@brendandburns wondering if there's anything I can do to help with reviewing this PR. i understand that this is a large chunk of code I created, so I would be happy to jump on a zoom call if you need a walkthrough.

brendandburns commented:
Yeah I need to find an hour or two in my schedule to review the code :)

I will try to block that out tomorrow

--brendan

________________________________
From: Andrew Stakhov <notifications@github.com>
Sent: Tuesday, April 7, 2020 2:05:41 PM
To: kubernetes-client/csharp <csharp@noreply.github.com>
Cc: Brendan Burns <bburns@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [kubernetes-client/csharp] Add support for informers (#394)


@brendandburns<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbrendandburns&data=02%7C01%7Cbburns%40microsoft.com%7C21966c9aeb9246f1414208d7db376e98%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637218903438875611&sdata=rL0miaJZM7XLPgl5MoVIyDeMl52IHf%2Fv3PQPPeHI0Fo%3D&reserved=0> wondering if there's anything I can do to help with reviewing this PR. i understand that this is a large chunk of code I created, so I would be happy to jump on a zoom call if you need a walkthrough.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkubernetes-client%2Fcsharp%2Fpull%2F394%23issuecomment-610620219&data=02%7C01%7Cbburns%40microsoft.com%7C21966c9aeb9246f1414208d7db376e98%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637218903438875611&sdata=h0DRbthxbi1MDYfwH6oVpePe%2BeamdTSenJBfr1fcg9c%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABL4HAQIWQSEFTPANATIOMDRLOISLANCNFSM4LYIPYWA&data=02%7C01%7Cbburns%40microsoft.com%7C21966c9aeb9246f1414208d7db376e98%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637218903438885611&sdata=Br%2F9xC%2FldaTucjjmb2K6pUERSatCezJ6btReq5gk3HI%3D&reserved=0>.

tg123 commented:
I am doing it too, but has to do after work :)

brendandburns commented on .github/workflows/dotnet.yaml:
```diff
@@ -8,32 +8,32 @@ jobs:
     name: Dotnet build
     steps:
-      - uses: actions/checkout@v1
+      - uses: actions/checkout@master
```
It's probably dangerous to track master because it could change. Is there a version that would work?

macsux replied:
This was switched from master to v1 2 days ago as part of 944622dd65abfd1c5475e1e1463f6624c626a5c2. I'm confused why, as wouldn't you want build & integration tests to run off master?

andykernahan replied:
That specifies the version of the action; it has no relation to the ref that is checked out.

This PR should be rebased and all build related commits removed.

andykernahan replied:
Apologies, I see that it has been rebased.

The action should not track master as it may eventually break the SxS workaround.

brendandburns commented on .github/workflows/dotnet.yaml:
```diff
@@ -8,32 +8,32 @@ jobs:
         uses: actions/setup-dotnet@v1
         with:
-          dotnet-version: 2.1.805
+          dotnet-version: 2.1.300
```
Why move backward in version-space?

macsux replied:
We actually have an indirect dependency on 2.1.0 in tests project which started throwing when strong signing was enabled on test project.

andykernahan replied:
@macsux 2.1.805 is the latest 2.1 SDK release and it would be preferable to use the bundled runtime.

> when strong signing was enabled on test project.

Can you elaborate on the error? Is it necessary to sign the tests?

macsux replied:
Strong signing for tests was enabled to allow unit testing internal classes/methods. In order to make test assembly "friend", it had to be strongly signed because main assembly is strongly typed. I don't remember the error now, but it was looking for specific version of some package that come with .net core 2.1.0. It was only showing up on the build server as I have every version of .NET on my dev machine.

andykernahan replied:
Understood (I had yet to reach that change).

I see the [build](https://github.com/kubernetes-client/csharp/actions/runs/73778750) for 2.1.805 succeed anyway.

macsux replied:
Ok, I reverted it to your version, not sure why but it works now. Gonna write off the error on something weird I was doing at the time.

brendandburns commented on examples/informers/ControllerService.cs:
```diff
@@ -0,0 +1,27 @@
+            var controllers = _serviceProvider.GetServices<IController>();
+            await Task.WhenAll(controllers.Select(x => x.Initialize(stoppingToken)));
+        }
+
```
delete empty lines.

brendandburns commented on examples/informers/ControllerService.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace informers
+{
+    public class ControllerService : BackgroundService
```
Document what this is an example of in a comment?

brendandburns commented on examples/informers/DeltaChangesQueryingController.cs:
```diff
@@ -0,0 +1,111 @@
+        private readonly ILogger _reconcilerLogger;
+        private ILogger _informerLogger;
+        CompareLogic _objectCompare = new CompareLogic();
+        // private readonly ActionBlock<List<ResourceEvent<V1Pod>>> _reconciler;
```
delete dead code.

brendandburns replied:
or add documentation if this is related to the example below. (or consider having two different examples)

brendandburns commented on examples/informers/DeltaChangesQueryingController.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            var workerQueue = _podInformer
+                .GetResource(ResourceStreamType.ListWatch, new KubernetesInformerOptions() { Namespace = "default"})
+                // .Resync(TimeSpan.FromSeconds(5))
```
why is this commented out?

brendandburns commented on examples/informers/DeltaChangesQueryingController.cs:
```diff
@@ -0,0 +1,111 @@
+
+            var obj = changes.First().Value;
+            var sb = new StringBuilder();
+            // sb.AppendLine($"Received changes for object with ID {KubernetesObject.KeySelector(obj)} with {changes.Count} items");
```
why is this commented out?

brendandburns commented on examples/informers/DeltaChangesQueryingController.cs:
```diff
@@ -0,0 +1,111 @@
+
+    // try creating and deleting some pods in "default" namespace and watch the output
+
+    public class DeltaChangesQueryingController : IController
```
This example has a lot of commented out code, I wonder if it might be better to just have two different files for the two different examples

brendandburns commented on src/KubernetesClient/Authentication/GcpTokenProvider.cs:
```diff
@@ -0,0 +1,57 @@
+
+namespace k8s.Authentication
+{
+    public class GcpTokenProvider : ITokenProvider
```
I assume this will get removed when your other PR merges.

macsux replied:
The PR is rebased on top of my other PR so it shows up here. I'll rebase again once the other PR goes in (which I assume will be the first one to be accepted given its smaller scope)

brendandburns commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,297 @@
+	{
+		private readonly Func<TResource, TKey> _keyFunc;
+		private readonly bool _skipTransient;
+        private readonly ILogger _log;
```
indentation is weird here.

brendandburns commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,297 @@
+using System;
```
Indentation is weird throughout this file, looks like it needs formatting?

brendandburns replied:
lack of formatting makes this file really hard to review.

brendandburns commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,297 @@
+
+			deltas.Add(obj);
+			CombineDeltas(deltas);
+
```
delete newlines.

brendandburns commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,297 @@
+		private readonly List<TargetLink> _targets = new List<TargetLink>();
+		private bool _isCompleting;
+		private long _msgId;
+		private readonly object _lock = new object();
```
Please document what this lock protects, it will make it easier to review/understand the code.

brendandburns commented on src/KubernetesClient/Extensions.cs:
```diff
@@ -0,0 +1,214 @@
+            source.Remove(key);
+            return true;
+        }
+        /// <summary>
```
newline between end of function and start of next function.

brendandburns commented on src/KubernetesClient/Extensions.cs:
```diff
@@ -0,0 +1,214 @@
+            {
+                result = queue.Dequeue();
+            }
+            catch (InvalidOperationException)
```
given that this only occurs if the queue is empty (and you test for that above) isn't this redundant?

macsux replied:
There is no lock so there's a chance that Count became zero when we actually execute Dequeue (which is why it's done in catch block). I'm avoiding expensive throw if queue is empty, but still protecting against race condition.

andykernahan replied:
`Queue<T>` is fundamentally not thread-safe: it must not be shared by multiple threads without proper serialization. If you have observed an `InvalidOperationException` being thrown, it would suggest improper sharing and/or locking.

macsux replied:
Removed catch block and made it internal.

andykernahan replied:
`TryPeek` has been updated, but the issue remains for `TryDequeue`.

macsux replied:
Should be fixed now

brendandburns commented on src/KubernetesClient/Extensions.cs:
```diff
@@ -0,0 +1,214 @@
+        /// <param name="source">The source enumerable</param>
+        /// <typeparam name="T">The type of elements</typeparam>
+        /// <returns>The produced hashset</returns>
+        public static HashSet<T> ToHashSet<T>(this IEnumerable<T> source)
```
Does this need to be a separate method?

macsux replied:
The method is used. It is present in .NET Core, but not in netstandard or net4x. It acts as polyfil.

andykernahan replied:
It would be preferable if all extensions methods are made internal except those which extend types declared by this project; otherwise, it simply pollutes the developer's completion list. `ToHashSet` being a case in point as it has been available since `netcoreapp2.0` and `net472`.

brendandburns commented on src/KubernetesClient/Informers/Cache/CacheSynchronized.cs:
```diff
@@ -0,0 +1,26 @@
+        /// Message number in the sequencer
+        /// </summary>
+        public long MessageNumber { get; }
+        /// <summary>
```
newlines between methods for readability.

brendandburns commented on src/KubernetesClient/Informers/Cache/ICache.cs:
```diff
@@ -0,0 +1,37 @@
+        /// Current version of cache
+        /// </summary>
+        long Version { get; set; }
+        /// <summary>
```
newlines (I'm going to stop commenting on this, but please apply the comment everywhere)

brendandburns commented on src/KubernetesClient/Informers/Cache/SimpleCache.cs:
```diff
@@ -0,0 +1,199 @@
+            }
+        }
+
+        public object SyncRoot { get; } = new object();
```
Move this to the beginning of the object, document that it is used as a lock.

brendandburns replied:
Also, why is it accessble if it is just used a lock?

macsux replied:
I originally introduced SyncRoot as ways to "pause" cache updates while doing some UOW but then moved towards Version & Snapshot based interface. It's not needed anymore, I'll remove

brendandburns commented on src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs:
```diff
@@ -0,0 +1,294 @@
+                }
+            }
+
+            public object SyncRoot => _parent._lock;
```
Do you do anything with this variable?

macsux replied:
Part of the original interface that later changed to use Version as synchronization mechanism. Removed

brendandburns commented on src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs:
```diff
@@ -0,0 +1,294 @@
+            public object SyncRoot => _parent._lock;
+            public ICacheSnapshot<TKey, TResource> Snapshot()
+            {
+                lock (_parent)
```
Shouldn't this be `_parent._lock`? (if not, document why not)

brendandburns replied:
Also, why are you locking around a constructor? Shouldn't that be thread-safe?

macsux replied:
1. Bug: changed to _parent._lock
2. The lock is there because snapshot uses current object as a  source dictionary (which gets copied in the constructor). This copying is not thread safe.

brendandburns commented on src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs:
```diff
@@ -0,0 +1,294 @@
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
```
Isn't this redundant with the method above?

macsux replied:
Part of interface of IDictionary, and no they are different.

brendandburns commented on src/KubernetesClient/Informers/Cache/VersionPartitionedSharedCache.cs:
```diff
@@ -0,0 +1,294 @@
+        // internal to allow for unit testing
+        internal readonly Dictionary<VersionResourceKey,TResource> Items = new Dictionary<VersionResourceKey, TResource>();
+        private readonly HashSet<CacheView> _views = new HashSet<CacheView>();
+        private readonly object _lock = new object();
```
All the lock semantics in this class are tricky, would it be better to use a concurrent collection?

macsux replied:
The reason locks are used and whole main reason for icache is to support atomic Reset operation.

brendandburns commented:
I did an initial round of review on the first 20 files or so. Will try to do more later.

If there's a way to break this PR up into smaller PRs (e.g Interfaces first, then implementation, then examples) it will make it much easier to review.

Also, formatting is off in places which makes it harder to review, please make sure to consistently format the code for easier reviewing.

macsux commented:
I've pushed a new version where I ran all my code through Rider code cleanup, it should fix all the spacing issues. I think because my tab and space were the same size everything lined up on my machine. 

I want to avoid splitting it into multiple PRs because I wanted to discuss the whole approach of whether it makes sense to keep adding stuff to the library as a single nuget package. I've commented on this as part of PR description, I feel like some of the dependencies introduced (or will be introduced as we progress further into controllers/operators) warrants making this project modular with multi-nuget packages. Much of the code that is currently part of informers example should actually be part of core library such as `.AddKubernetes`, but it drags dependencies that not every project needs. 

Given that informers drags in RX.NET, AsyncEnumerables, imo it should be isolated as own nuget package that depends on KubernetesClient nuget. Controllers drags in TPL, so imo those building blocks should be further isolated building on top of informers. I plan on adding a lot more in that space going forward.

macsux commented:
Strong signing for tests was enabled to allow unit testing internal
classes/methods. In order to make test assembly "friend", it had to be
strongly signed because main assembly is strongly typed. I don't remember
the error now, but it was looking for specific version of some package that
come with .net core 2.1.0. It was only showing up on the build server as I
have every version of .NET on my dev machine.

On Wed, Apr 8, 2020 at 1:47 PM Andy Kernahan <notifications@github.com>
wrote:

> *@andykernahan* commented on this pull request.
> ------------------------------
>
> In .github/workflows/dotnet.yaml
> <https://github.com/kubernetes-client/csharp/pull/394#discussion_r405703436>
> :
>
> >          uses: actions/setup-dotnet@v1
>          with:
> -          dotnet-version: 2.1.805
> +          dotnet-version: 2.1.300
>
> @macsux <https://github.com/macsux> 2.1.805 is the latest 2.1 SDK release
> and it would be preferable to use the bundled runtime.
>
> when strong signing was enabled on test project.
>
> Can you elaborate on the error? Is it necessary to sign the tests?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-client/csharp/pull/394#discussion_r405703436>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINFWEJKK4QCZ6MPQRRNJLRLS2BVANCNFSM4LYIPYWA>
> .
>

andykernahan commented on examples/informers/informers.csproj:
```diff
@@ -0,0 +1,20 @@
+        <PackageReference Include="CompareNETObjects" Version="4.65.0" />
+        <PackageReference Include="Microsoft.Extensions.Hosting" Version="3.1.2" />
+        <PackageReference Include="Microsoft.Extensions.Http" Version="3.1.2" />
+        <PackageReference Include="Microsoft.Extensions.Http" Version="3.1.2" />
```
Duplicate.

macsux replied:
Fixed

andykernahan commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,284 @@
+        public void Fault(Exception exception)
+        {
+            _taskCompletionSource.SetException(exception);
+            foreach (var link in _targets.Where(x => x.LinkOptions.PropagateCompletion)) link.Target.Fault(exception);
```
The enumeration must be done with the lock held; otherwise, it's possible an `InvalidOperationException` will be thrown if the collection is mutated concurrently.

andykernahan commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,284 @@
+                    _targets.Insert(0, link);
+
+                OfferMessageToLink(link);
+                return Disposable.Create(() => _targets.Remove(link));
```
The removal must be done with the lock held.

andykernahan commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,284 @@
+            _log.LogTrace($"Link replied with {link.LastOfferedMessageReply} after being offered {string.Join(string.Empty, msg)}");
+            if (link.LastOfferedMessageReply == DataflowMessageStatus.DecliningPermanently)
+            {
+                _targets.Remove(link);
```
The removal must be done with the lock held.

andykernahan commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,284 @@
+                }
+            }
+
+            var header = new DataflowMessageHeader(++_msgId);
```
`++_msgId` is not atomic.

andykernahan commented on src/KubernetesClient/Informers/FaultTolerance/RetryPolicy.cs:
```diff
@@ -0,0 +1,76 @@
+            while (true)
+                try
+                {
+                    return await action();
```
All awaits should be configured with`ConfigureAwait(false)`.

andykernahan commented on src/KubernetesClient/Informers/FaultTolerance/RetryPolicy.cs:
```diff
@@ -0,0 +1,76 @@
+                    if (!ShouldRetry(e, retryCount))
+                        throw;
+                    retryCount++;
+                    await Task.Delay(RetryDelay(retryCount));
```
`ConfigureAwait(false)`

andykernahan commented on src/KubernetesClient/Informers/KubernetesInformer.cs:
```diff
@@ -0,0 +1,155 @@
+            public IObservable<ResourceEvent<TResource>> GetObservable()
+            {
+                var result = Observable.Empty<ResourceEvent<TResource>>();
+                if (_type != ResourceStreamType.Watch) result = result.Concat(List());
```
The use of `HasFlag` would be clearer. Same for line beneath.

andykernahan commented on src/KubernetesClient/Informers/KubernetesInformerOptions.cs:
```diff
@@ -0,0 +1,17 @@
+        /// <summary>
+        ///     The default options for kubernetes informer, without any server side filters
+        /// </summary>
+        public static KubernetesInformerOptions Default { get; } = new KubernetesInformerOptions();
```
Given this type is not immutable, it's dangerous to expose a singleton; `Default.Namespace` could be accidentally mutated, thus changing the behaviour of the default shared informers.

Consider making it immutable and: 
1.  Add a ctor with optional arguments. There is precedence for this pattern, but it does not version well (i.e. adding a parameter is a breaking chance).
2. Add a fluent build. There is no precedence for this pattern, so it's worth gathering the thoughts of other's beforehand (@tg123 , @brendandburns).

macsux replied:
I've created a fluent builder, as I was planning to do it anyways for labels. Unit test coverage for builder is also added.

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,294 @@
+
+        public static IObservable<TOut> WithResets<TResource, TOut>(this IObservable<ResourceEvent<TResource>> source, Func<List<ResourceEvent<TResource>>, IEnumerable<ResourceEvent<TResource>>> resetSelector, Func<ResourceEvent<TResource>, TOut> itemSelector)
+        {
+            var resetBuffer = new List<ResourceEvent<TResource>>();
```
`resetBuffer` will be shared across all subscriptions and I do not believe that is correct.

The declaration should be moved into the lambda supplied to `Create`.

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,294 @@
+
+                            observer.OnNext(itemSelector(notification));
+                        }, OnError, OnComplete);
+                    return new CompositeDisposable(upstreamSubscription, Disposable.Create(OnComplete));
```
Consider `StableCompositeDisposable.Create` instead to create disposables from a fixed number of disposables.

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,294 @@
+        {
+            if (mapper == null)
+                mapper = (oldResource, newResource) => newResource;
+            long msgNum = 0;
```
`msgNum` will be shared across all subscriptions and I do not believe that is correct.

Consider using `Observable.Defer` to close over `msgNum` instead.

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,294 @@
+
+        public static IObservable<ResourceEvent<TResource>> ComputeMissedEventsBetweenResets<TKey, TResource>(this IObservable<ResourceEvent<TResource>> source, Func<TResource, TKey> keySelector, IEqualityComparer<TResource> comparer)
+        {
+            Dictionary<TKey, TResource> cacheSnapshot;
```
These variables will be shared across all subscriptions and I do not believe that is correct.

Their declarations should be moved into the lambda supplied to Create.

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,294 @@
+                        observer.OnCompleted();
+                        timerSubscription.Dispose();
+                    });
+                return new CompositeDisposable(timerSubscription, sourceSubscription);
```
Consider `StableCompositeDisposable.Create`.

macsux replied:
Changed, but can you please explain the difference between these two? I read the description of the class, but I still don't understand what the difference is from a functional perspective.

andykernahan replied:
`CompositeCollection` is intended for scenarios where the number of disposables is not fixed; they can be arbitrarily added, removed, cleared etc.

`StableCompositeDisposable` is intended for scenarios where the number is fixed. `Create` in this instance [returns](https://github.com/dotnet/reactive/blob/08b377703c1976601c63c93871450738c00898bb/Rx.NET/Source/src/System.Reactive/Disposables/StableCompositeDisposable.cs#L33) an implementation specialised for two disposables, thereby eliminating several array allocations when compared to `CompositeCollection`.

andykernahan commented on src/KubernetesClient/Informers/ResourceStreamType.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <summary>
+        ///     A Hot observable that Lists current state of resources followed by watch.
+        /// </summary>
+        ListWatch = 3
```
Consider `ListWatch = List | Watch` for clarity.

macsux replied:
I originally had it that way, however, given that this is public API that developers will use commonly I was trying to avoid forcing people to use bitvise combinations on flags. A lot of developers are actually not familiar with Flags and how to properly use them. ListWatch gives a "simple" option for them to use that is intuitive. This is unlike `EventTypeFlags` which is considered "low level" API and most developers are unlikely to interact with it directly, but even if they do I made sure the naming convention is very clear to that they should be apply HasFlag when using it.

andykernahan replied:
I simply meant changing the declaration of the enum member to be:
```csharp
public enum ResourceStreamType
{
    ListWatch = List | Watch
}

```

andykernahan commented on src/KubernetesClient/Informers/SharedInformer.cs:
```diff
@@ -0,0 +1,187 @@
+        private TaskCompletionSource<bool> _cacheSynchronized = new TaskCompletionSource<bool>();
+        private readonly IConnectableObservable<CacheSynchronized<ResourceEvent<TResource>>> _masterObservable;
+
+        private IScheduler _masterProcessScheduler = new EventLoopScheduler();
```
This is not used. Is it supposed to be?

andykernahan commented on src/KubernetesClient/Informers/SharedInformer.cs:
```diff
@@ -0,0 +1,184 @@
+                    AddSubscriber();
+                    _logger.LogTrace("Subscriber awaiting cache synchronization before attaching");
+
+                    var isCacheSynchronized = await _cacheSynchronized.Task;
```
`ConfigureAwait(false)`

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,299 @@
+                            .Union(updatedEntities)
+                            .Union(resetEntities);
+                    })
+                    .Do(_ => Console.WriteLine(_))
```
It would be preferable if we didn't spam the console.

macsux replied:
Accidentally left when debugging (before I introduced logger). Will remove

andykernahan commented on src/KubernetesClient/Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,299 @@
+                    {
+                        if (!cacheSynchronized) return resetBuffer;
+
+                        cacheSnapshot = new Dictionary<TKey, TResource>(cache);
```
Consider moving `cacheSnapshot` declaration here; otherwise, it's unnecessarily rooted for the lifetime of the subscription.

andykernahan replied:
Can this be `cache.Snapshot()` instead?

macsux replied:
Good catch. This code was written before I introduced Snapshot as part of ICache.

andykernahan commented on src/KubernetesClient/Informers/SharedInformer.cs:
```diff
@@ -0,0 +1,184 @@
+            _logger.LogTrace("Finished yielding to subscribers");
+        }
+
+        private void AddSubscriber()
```
The locking in `AddSubscriber` and `RemoveSubscriber` must be reconsidered given subscribers `SubscribeOn` and `ObserveOn` their own `EventLoopScheduler`, meaning the methods will be invoked concurrently. In particular, access to `_subscribers` must be serialised.

andykernahan commented on src/KubernetesClient/Informers/SharedInformer.cs:
```diff
@@ -0,0 +1,184 @@
+                    return Observable.Create<ResourceEvent<TResource>>(observer =>
+                        {
+                            var broadcasterAttachment = Disposable.Empty;
+                            var cacheSnapshot = _cache.Snapshot();
```
Is it necessary to snapshot the cache? It seems we only need to capture the version of the cache.

andykernahan commented on src/KubernetesClient/Informers/Cache/SimpleCache.cs:
```diff
@@ -0,0 +1,196 @@
+
+namespace k8s.Informers.Cache
+{
+    public class SimpleCache<TKey, TResource> : ICache<TKey, TResource>, ICacheSnapshot<TKey, TResource>
```
I do not believe the cache should implement `ICacheSnapshot` as it does not satisfy the contact of a "snapshot of cache at a point in time".

Consider:
```csharp
ICache<int, int> cache = new SimpleCache<int, int>();
ICacheSnapshot<int, int> snapshot = cache;
cache.Add(42, 0);
snapshot.ContainsKey(42); // returns true when it should be false.
```

macsux replied:
This would be a weird way of working with this class, though from purist perspective you're right. I was trying to reuse the same object for both snapshotting and "live" (at least for SimpleCache implementation). I guess I can create a some kind of childview type object for SimpleCache snapshot. Wondering if others have opinion on this before I fix: @brendandburns @tg123 ?

andykernahan replied:
I would be hesitate to make assumptions about how a type is to be used, especially one that is public and implicitly convertible to an interface that it does not implement correctly.

Here's a Rider generated implementation:
```csharp
private sealed class SimpleCacheSnapshot : ICacheSnapshot<TKey, TResource>
{
    private readonly Dictionary<TKey, TResource> _items;
    public SimpleCacheSnapshot(SimpleCache<TKey,TResource> cache, long version)
    {
        _items = new Dictionary<TKey, TResource>(cache._items);
        Version = version;
    }
    public IEnumerator<KeyValuePair<TKey, TResource>> GetEnumerator() => _items.GetEnumerator();
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    public int Count => _items.Count;
    public bool ContainsKey(TKey key) => _items.ContainsKey(key);
    public bool TryGetValue(TKey key, out TResource value) => _items.TryGetValue(key, out value);
    public TResource this[TKey key] => _items[key];
    public IEnumerable<TKey> Keys => _items.Keys;
    public IEnumerable<TResource> Values => _items.Values;
    public long Version { get; }
}
```

macsux replied:
Ok, changed as recommended

andykernahan commented on src/KubernetesClient/Informers/Cache/SimpleCache.cs:
```diff
@@ -0,0 +1,196 @@
+            Version = version;
+            _items = new Dictionary<TKey, TResource>(items);
+        }
+
```
Nit: redundant blank lines in a couple of places.

andykernahan commented on src/KubernetesClient/Informers/SharedOptionsInformer.cs:
```diff
@@ -0,0 +1,39 @@
+    /// </summary>
+    /// <typeparam name="TResource"></typeparam>
+    /// <typeparam name="TOptions"></typeparam>
+    public class SharedOptionsInformer<TResource, TOptions> : IInformer<TResource, TOptions>
```
I presume this type is supposed to be thread-safe? It's assumed to be so in the informer example as (the derived) `SharedKubernetesInformer<,>` is registered as a singleton.

andykernahan commented on src/KubernetesClient/Informers/SharedInformer.cs:
```diff
@@ -0,0 +1,184 @@
+                                    // we could be ahead of broadcaster because we initialized from cache which gets updated before the message are sent to broadcaster
+                                    // this logic realigns us at the correct point with the broadcaster
+                                    .Do(x => _logger.LogTrace($"Received from broadcaster {x}"))
+                                    .SkipWhile(x => x.MessageNumber <= cacheSnapshot.Version)
```
Can you elaborate why this is `MessageNumber` instead of `CacheVersion`?

macsux replied:
What ends up happening is when reset completes, cache gets synchronized up to the LAST message is reset block, however the actual messages haven't been "broadcast" yet. The result is that cache version is actually AHEAD of what is being sent out of the broadcaster. When new subscriber joins, they "List" from cache, and their position in the event stream can very well end up ahead of the broadcaster. If they are trying to watch after, they need to skip any messages that were already included in the cache and only start processing them once the version of cache they listed from is equals to what is being sent out of broadcaster. This is the whole reason for CacheSynchronized datastructure as I struggled with this synchronization dilemma for a long time and finally settled on this approach.

andykernahan commented on src/KubernetesClient/Informers/SharedInformer.cs:
```diff
@@ -0,0 +1,184 @@
+
+        public IObservable<ResourceEvent<TResource>> GetResource(ResourceStreamType type)
+        {
+            var childScheduler = new EventLoopScheduler(); // dedicated thread for the child on which all messages are syncrhonized
```
The scheduler must be disposed; otherwise, its thread [waits](https://github.com/dotnet/reactive/blob/e00d53bd022dd2a787d983e0fbbcd04016978bc2/Rx.NET/Source/src/System.Reactive/Concurrency/EventLoopScheduler.cs#L298) indefinitely for an event that will never be signalled.

macsux replied:
I'm looking for a little help on how to do this properly. Adding it as part of observable `Dispose` lifeline terminates it too early and breaks the whole thing. A bit of a dirty hack would be to put it on something like Task.Delay(5000) to allow the stream to finalize and close before the actual thread is killed off. Let me know if you have a better idea.

andykernahan replied:
Might `Observable.Using` help?

macsux replied:
That did it. I was not aware of that operator.

andykernahan commented on src/KubernetesClient/Informers/SharedKubernetesInformer.cs:
```diff
@@ -0,0 +1,46 @@
+    /// </summary>
+    /// <typeparam name="TResource">The type of resource to monitor</typeparam>
+    public class SharedKubernetesInformer<TResource> :
+        SharedOptionsInformer<TResource, KubernetesInformerOptions>,
```
This would benefit from `KubernetesInformerOptions` being made immutable and equatable; currently, a new `SharedInformer<,>` will be created for options that are considered equal.

macsux replied:
That was always the intent to allow subscribers that have the same options to share the same shared informer (which is locked to a single type of options). I could have swore I implemented equality operators. Thanks for spotting this.

andykernahan commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,299 @@
+            }
+
+            var header = new DataflowMessageHeader(Interlocked.Increment(ref _msgId));
+            link.LastOfferedMessageReply = link.Target.OfferMessage(header, msg, this, true);
```
Is it safe to mutate `link` without the lock held? This method can be invoked concurrently.

andykernahan commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,299 @@
+            foreach (var link in linksToOfferTo)
+            {
+                DataflowMessageStatus reply;
+                do // keep feeding the link messages until queue is either empty or it tells us that it can't handle any more
```
Might the developer not expect the block of have broadcast semantics? For example:
```csharp
var deltas = informer.ToResourceEventDeltaBlock(...);
deltas.LinkTo(new ActionBlock<...>(Log));
deltas.LinkTo(new ActionBlock<...>(BuildState));
```
I would expect `Log` and `BuildState` to process all deltas.

macsux replied:
From TPL perspective, there's no reason for them to expect broadcast behavior as messages have to be explicitly routed through a Broadcast block and "cloned" as access to the message is supposed to be thread-safe in Actor programming. None of the other TPL blocks exhibit implicit broadcast semantics. If they wanted to do broadcast they can always do `.Publish().AutoConnect()` prior to calling `ToResourceEventDeltaBlock`.

andykernahan commented on src/KubernetesClient/Informers/KubernetesInformer.cs:
```diff
@@ -0,0 +1,155 @@
+                                        return new[] {x.Object.ToResourceEvent(EventTypeFlags.Modify)};
+                                    case WatchEventType.Bookmark:
+                                        if (x.Object is IMetadata<V1ObjectMeta> status)
+                                            _resourceVersion = status.Metadata.ResourceVersion;
```
I'm not sure that bookmarks can be solely replied upon to keep `_resourceVersion` up to date. The docs state:
>  Clients should not assume bookmarks are returned at any specific interval, nor may they assume the server will send any BOOKMARK event during a session.

We should use the same approach as the Go [client](https://github.com/kubernetes/client-go/blob/2a5c88adcc5c656a3e843a3f2ca4633cdb9b4227/tools/cache/reflector.go#L474) and update `_resourceVersion` on every `{Added, Deleted, Modified, Bookmark}` event.

macsux replied:
So this part has me really confused and I couldn't find a straight forward answer. Conceptually we have resourceVersion on LIST object and each item in it. When you execute a List or Watch, you're actually supplying a resourceVersion of the LIST. The relationship between list version and individual item versions is not described anywhere. For example if they were implemented as a simple integer increments, then LIST version would have completely different semantics from items in it, and its own counter. On the other hand if we're saying that it's just a timestamp of latest mutation of the list, then obtaining the latest "version" from a Add/Deleted/Modified resource can be applied to List version. I could not find any documentation in K8S to clarify if this assumption is safe to make given that they state resourceVersion should be treated "opaque". If go does this, then I'll apply same behavior, I just wanted someone to clarify this for me.

macsux replied:
I've changed it to take a resource version from any watch event that has it set, not just bookmark. Double-check please that this is what we're expecting.

tg123 commented on examples/informers/Program.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        public static void Main(string[] args)
+        {
+            CreateHostBuilder(args).Build().Run();
```
can this example be a non asp.net app
I mean pure console app

macsux replied:
This is not an aspnet - it's `Microsoft.NET.Sdk.Worker`. The included example is of a simple controller, and as such is expected to work as a worker code. The informer code is also intended to be used primarily with DI.

macsux commented on .dockerignore:
```diff
@@ -0,0 +1,6 @@
+**/bin/
```
The sample project uses NMica to autogenerate dockerfile on build and dockerignore if it's missing. Both can actually be added to .gitignore since they are autogenerated

macsux commented on .gitignore:
```diff
@@ -13,3 +13,4 @@ bin/
 # JetBrains Rider
 .idea/
 *.sln.iml
+examples/informers/Dockerfile
```
Autogenerated on build

macsux commented on src/KubernetesClient/KubernetesClient.csproj:
```diff
@@ -8,37 +8,41 @@
-
-    <LangVersion>7.3</LangVersion>
-    <TargetFrameworks>netstandard2.0;net452;netcoreapp2.1</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;net461;netcoreapp2.1</TargetFrameworks>
```
The entire informers is built on System.Reactive which is netstandard2.0. net461 is the earliest framework that supports netstandard2.0 spec. 

I flagged this in PR description at the top. I actually think that instead of having a single nuget package we should modularize it. Informers can be something like KubernetesClient.Informers. The controllers same thing as it built on top of informers. This also goes towards needing to have a place for `.AddKubernetes` DI method which is currently part of sample project, but that's not where it belongs - should be part of the SDK.

macsux commented on src/KubernetesClient/KubernetesClient.csproj:
```diff
@@ -8,37 +8,41 @@
     <PackageReference Include="Fractions" Version="4.0.1" />
     <PackageReference Include="Microsoft.AspNetCore.JsonPatch" Version="1.1.2" Condition="'$(TargetFramework)' != 'netstandard2.0' and '$(TargetFramework)' != 'netcoreapp2.1'" />
     <PackageReference Include="Microsoft.AspNetCore.JsonPatch" Version="3.0.0" Condition="'$(TargetFramework)' == 'netstandard2.0' or '$(TargetFramework)' == 'netcoreapp2.1'" />
+    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="2.0.0" />
```
If I get an agreement from maintainers to introduce additional projects as part of SDK umbrella, I'm all for it. @brendandburns are you ok with doing this?

macsux commented on src/KubernetesClient/KubernetesClient.csproj:
```diff
@@ -8,37 +8,41 @@
     <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0" PrivateAssets="All" />
+    <PackageReference Include="System.Reactive" Version="4.3.2" />
+    <PackageReference Include="System.Interactive.Async" Version="4.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.2.0" />
```
Informers are meant to be used in DI context. I only took on dependency on abstractions here to have the interfaces, not the actual implementations

macsux commented on tests/KubernetesClient.Tests/KubernetesClient.Tests.csproj:
```diff
@@ -1,23 +1,35 @@
     <TargetFrameworks>netcoreapp2.1;netcoreapp2.0</TargetFrameworks>
+    <LangVersion>8</LangVersion>
+      <SignAssembly>true</SignAssembly>
+      <AssemblyOriginatorKeyFile>..\..\src\KubernetesClient\kubernetes-client.snk</AssemblyOriginatorKeyFile>
```
I'm unit testing internal methods / classes, so needed to make test project friend of main assembly. But can't do that without signing test project since main is signed

macsux commented on src/KubernetesClient/Informers/Cache/SimpleCache.cs:
```diff
@@ -0,0 +1,208 @@
+    public class SimpleCache<TKey, TResource> : ICache<TKey, TResource>
+    {
+        private readonly IDictionary<TKey, TResource> _items;
+        private readonly object _syncRoot = new object();
```
Due to semantics around needing to enforce Reset operation as atomic, the lock has to be used against every mutation of the dictionary. Concurrent dictionary does not expose atomic "Replace" semantics. This is the main reason why ICache exists in first place.

macsux commented on src/KubernetesClient/Informers/FaultTolerance/Extensions.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace k8s.Informers.FaultTolerance
+{
+    public static class Extensions
```
I wanted to discuss, as it's possible someone may want to override fault tolerance behavior. I tried using Polly, but it doesn't work with RX.

macsux commented:
@tg123 @brendandburns  before splitting it up into multiple PRs, I would like to get a consensus on the direction of the whole thing, primarily whether we're ok with putting some of this stuff into a separate nuget library. As I've mentioned before, RX.NET requires netstandard2.0 so it would require bumping up to 461 at the very minimum. Additionally, DI specific extensions are generally provided as their own specialized packages given that not everyone will be using built in DI framework, but most people will and it's good to have helper registration methods available to them. This can be seen as part of the sample project, but this code does not belong there.

davidfowl commented on examples/informers/ServiceCollectionKubernetesExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+        public static IServiceCollection AddKubernetes(this IServiceCollection services)
+        {
+            services.AddHostedService<ControllerService>();
+            var controllers = AppDomain.CurrentDomain.GetAssemblies()
```
I would refrain from any scanning like this being built in to core methods like AddKubernetes. 

This should also be split into its own package.

macsux replied:
This is what I suggested: modularize the SDK into multiple packages. I'm still waiting for a blessing from maintainers to do so before I put shovels in the ground

macsux replied:
I was trying to create similar pattern to what ASP.NET core uses, which also does assembly scanning to discover Controllers.

davidfowl replied:
@macsux Don't 😄

macsux replied:
Lol. Isn't asp.net your baby? I'm trying to make this as intuitive as possible for dotnet developer to write custom controllers, operators. I hate this status quo atm where kubernetes = write everything in Go. C# is just so elegant

davidfowl replied:
Yes but scanning assemblies is my arch-nemesis. MVC is the only thing that does it and I'm actively protesting against it 😄

macsux replied:
Ok, I've switched it to be explicit in the sample project. Stay tuned for some kickass sample .NET k8s operators /w blog posts built on top of all this work :)

davidfowl commented on examples/informers/ServiceCollectionKubernetesExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+            foreach (var controller in controllers) services.AddSingleton(typeof(IController), controller);
+            var config = KubernetesClientConfiguration.BuildDefaultConfig();
+
+            services.AddSingleton(config);
```
Does this need to be in the container? Seems like it would be a bit safer to have a custom options type here.

macsux replied:
Added it as parameter to `AddKubernetes`

davidfowl commented on examples/informers/ServiceCollectionKubernetesExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+            var config = KubernetesClientConfiguration.BuildDefaultConfig();
+
+            services.AddSingleton(config);
+            services.AddSingleton<RetryDelegatingHandler>();
```
Where is this used if it's being created further down manually?

macsux replied:
Left over from later refactoring. Removed now

davidfowl commented on src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs:
```diff
@@ -0,0 +1,304 @@
+    /// </summary>
+    /// <typeparam name="TKey">The type of key used to identity resource</typeparam>
+    /// <typeparam name="TResource">The resource type</typeparam>
+    public class ResourceEventDeltaBlock<TKey, TResource> : ISourceBlock<List<ResourceEvent<TResource>>>, ITargetBlock<ResourceEvent<TResource>>
```
Do we need a dependency on dataflow? This library is extremely allocation inefficient.

macsux replied:
I chose TPL dataflow because when you deal with controllers, it exposes useful queue like semantics which are intuitive. I could model this with RX, but to gain same functionality we're gonna need two observables one that is supplying data, and another to mark boundaries of when worker is ready for more. This will end up very similar in syntax to RX Window function, which to most dev not familiar with RX goes right over their head.

I brought this up a number of times, but I actually want to get blessing to modularize the SDK, and then TPL specific stuff around workflow scheduling for controllers would be moved there. I see a need for at least 3 extra packages:
KubernetesClient.Informers - all the RX stuff
KubernetesClient.Controllers - all the TPL stuff, a new pieces of code I plan to add to create reusable reconciliation plans
KubernetesClient.DependencyInjection - stuff to add all the extension methods into like the one currently in sample project

macsux replied:
I originally had it coded with System.Threading.Channels, but switched to TPL Dataflow as it gives a fine-grained control over "number of concurrent workers" semantics.

davidfowl replied:
I think you should switch it back.

macsux replied:
Removed this code for now, will reimplement on top of RX without TPL Dataflow once more design patterns for .NET operators are flushed out.

tg123 commented:
> @tg123 @brendandburns before splitting it up into multiple PRs, I would like to get a consensus on the direction of the whole thing, primarily whether we're ok with putting some of this stuff into a separate nuget library. As I've mentioned before, RX.NET requires netstandard2.0 so it would require bumping up to 461 at the very minimum. Additionally, DI specific extensions are generally provided as their own specialized packages given that not everyone will be using built in DI framework, but most people will and it's good to have helper registration methods available to them. This can be seen as part of the sample project, but this code does not belong there.

As a library, it is better to have as less dependency as possible. 
and should upgrade targeting versions (netstandard net452 or etc) carefully 

I wish we can drop support to `net452` but it is not a good idea when there are still many users.
or set a sunset date for `net452`

btw, we do not have to target `net461`, it already support `netstandard2.0`

macsux commented:
@tg123 
> As a library, it is better to have as less dependency as possible. I wish we can drop support to net452 but it is not a good idea when there are still many users.
or set a sunset date for net452

What about the idea of creating new packages that are part of the project as I've mentioned. I think this is the best approach.

macsux commented:
@tg123 @brendandburns Given that there is a desire to keep support for 452 in place and introduction of new dependencies, I've split this into two new packages: Informers and DependencyInjection. I've also isolated pre-reqs into separate PR https://github.com/kubernetes-client/csharp/pull/412.
I've also adjusted NBGV to now version each package independently.

Given that there was some scrutiny of use of TPL DataFlow from @davidfowl , I've taken that out as part of this PR and adjusted the sample to use a more primitive logic in the interim.

macsux commented on src/KubernetesClient/Kubernetes.ConfigInit.cs:
```diff
@@ -168,8 +168,12 @@ private void InitializeFromConfig(KubernetesClientConfiguration config)
             var cur = FirstMessageHandler as DelegatingHandler;
-
-            while (cur != null)
+            if (cur == null)
```
When using own HttpClient, it doesn't properly initialize this and will throw if not initialized like this

davidfowl commented on examples/informers/DeltaChangesQueryingController.cs:
```diff
@@ -0,0 +1,91 @@
+            return Task.CompletedTask;
+        }
+
+        private async void Reconcile(IList<ResourceEvent<V1Pod>> changes)
```
async void makes me sad. Also what happens when these calls overlap?

macsux replied:
Async is an artifact of refactoring, this can just be void for this sample. I've updated sample. Given that I took out the code you didn't like that was on top of TPL, this is very primitive controller and not something you would use in production. Worker / reconciler type manager is going to be part of controller  building blocks which is something I still need to tackle. Those new blocks would ensure that we're only running a single reconcile process against a given resource and potentially support cancellation semantics (no point in finishing current reconciliation if the desired state has changed). For now, this is meant to highlight the informers piece.

davidfowl commented on examples/informers/IController.cs:
```diff
@@ -0,0 +1,17 @@
+    /// <summary>
+    ///     Base interface for implementing controllers
+    /// </summary>
+    public interface IController
```
Feels like this should be a bit more prescriptive. It doesn't do much right now.

macsux replied:
This is not final interface and is meant to explore what controllers interface should look like. I'll probably extend it to support shutdown hook as well. This is in a sample project so it's not part of public API.

davidfowl replied:
I see I didn't catch that. Makes sense.

davidfowl commented on src/KubernetesClient.Informers/Extensions.cs:
```diff
@@ -0,0 +1,174 @@
+
+namespace k8s.Informers
+{
+    public static class Extensions
```
I'd make this internal.

macsux replied:
There are a few methods in there that are meant to be public, mainly `DisposeWith` and `CombineWith` to help in building up RX code. `Watch` is also useful for anyone that wants to consume raw Watcher output as observable. I've already switch `polyfill` type code (ToHashset, etc) to be internal.

anaisbetts commented on src/KubernetesClient.Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,273 @@
+                        .Do(x =>
+                        {
+                            if (!x.EventFlags.HasFlag(EventTypeFlags.Reset))
+                                throw new InvalidOperationException("Resync was applied to an observable sequence that does not issue a valid List event block when subscribed to");
```
Throwing in a `Do` block is probably Not Great, a better way would be to write:

```csharp
.SelectMany(x => x.EventFlags.HasFlag(EventTypeFlags.Reset) ? 
  Observable.Throw(new InvalidOperationException("...") :
  Observable.Return(x));
```

richbryant replied:
That's cleaner but `Do` tends to leave the user with no doubt what's happening.

Judgement call.  I prefer Ana's way but for the less Rx-minded, `Do` is fine.

macsux replied:
Unless there is a specific issue with the way the code is currently written, I think I'll leave at as `.Do`. I feel like the proposed method will actually be significantly heavier as it would allocate a new observable for EACH message.

anaisbetts commented on src/KubernetesClient.Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,273 @@
+            return Observable.Create<ResourceEvent<T>>(observer =>
+            {
+                var timerSubscription = Observable
+                    .Interval(timeSpan)
```
This needs a Scheduler parameter, probably should be an optional parameter of the Resync method

anaisbetts commented:
Threw in a few comments as a drive-by on @davidfowl's recommendation, thanks for all this great work @macsux!

mdekrey commented on src/KubernetesClient.Informers/ResourceObservableExtensions.cs:
```diff
@@ -0,0 +1,273 @@
+
+            return Observable.Defer(() =>
+            {
+                long msgNum = 0;
+
+                return source
+                    .Do(_ => msgNum++)
```
This `msgNum` being stored outside but mutated with a Do within the observable concerns me. With the purpose of it being simply a Message Number, I expect it's probably fine... but I can't help but think there's a better way.

macsux replied:
As long as closure semantics are well understood, this is safe to do. It's done in a few other places. RX itself uses a special pattern that is based on stateful classes to model new RX operators, but all this infrastructure is internal. I wanted to keep this light.

mdekrey replied:
Thanks for the explanation - it makes sense with this being in a new operator.

tg123 commented:
anyway to keep informer in the main package?
I do not think we could implement informer on net452

BTW, I do not want to support `net45` either. but it is widely used.
the code would be even more clean if dropping old netframework support.
or stop supporting `net452` when `3.0` (would be `k8s 1.9`) ? @brendandburns

macsux commented:
I think it's not a good idea to try to keep everything in one package.
This is consistent with core direction and moving away from monolithic
releases. RX requires netstandard20 so there is no way to keep it in main
package. There are more things coming down the pipe and new dependencies
being introduced where developer should choose what they need. For example
there is strong case to be made for DI extension methods, controller pieces
which will probably be core only.



On Wed., Apr. 22, 2020, 7:47 a.m. Boshi Lian, <notifications@github.com>
wrote:

> anyway to keep informer in the main package?
> I do not think we could implement informer on net452
>
> BTW, I do not want to support net45 either. but it is widely used.
> the code would be even more clean if dropping old netframework support.
> or stop supporting net452 when 3.0 (would be k8s 1.9) ? @brendandburns
> <https://github.com/brendandburns>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-client/csharp/pull/394#issuecomment-617727872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINFWCT47QBTBDOCRPAUDTRN3KLNANCNFSM4LYIPYWA>
> .
>

brendandburns commented:
Having this in a separate package makes sense to me. I'm going to try to allocate another hour or so to re-review this code once it gets rebased. It's a really big PR so it's challenging to find the time to give it an adequate review.

macsux commented:
I've recently got agreement from project steeltoe.io which I contribute to and is mainly maintained by efforts from my current company (VMWare) to take this on as part of their codebase. The way I structured informers makes the codebase high-level LINQ abstraction over resources that can be watched for changes, with Kubernetes being just one of the potentially many implementations. Similar relationship as EF has with underlying drivers. Given a relatively loose dependency on Kubernetes, unless there's a strong desire to keep it here, I'm going to move this chunk over to steeltoe.io codebase with dependency on Kubernetes client. 

The plan right now is to introduce two new packages as part of that codebase:
Steeltoe.Informers.InformersBase which will provide the core abstraction model
Steeltoe.Informers.KubernetesInformer which will provide Kubernetes specific implementation with dependency on KubernetesClient. 

That codebase is already setup to be multi-package so its easier to integrate, and easier to introduce large code changes like this because of how the release cycle is structured.

I would like to hear your thoughts before I close this PR.

brendandburns commented:
Obviously it's fine with me if you want to move this to a different package.

I think it's also possible to make it work within this repo and I'm happy to help get it there, but if you feel like the Steeltoe project is a better fit, then go for it.

fwiw, we will probably add informers to this package at some point, so we may look to the code in steeltoe for inspiration eventually.

macsux commented:
I wrote an extensive blog post about idea behind informer. @brendandburns if you wanna read through it and let me know your thoughts - it should give you an idea of what this is trying to accomplish. It would be good to leave Kubernetes specific informer piece in here (as a driver) for the base informer framework.

macsux commented:
Woops, forgot the link
https://stakhov.pro/informers-intro/

RichiCoder1 commented:
Would it be possible to revive this PR in some fashion?

macsux commented:
This code will added as part of project SteelToe.io over the next few
months. Main reason is that I'm using newer apis which prevent me from
contributing them here. Second reason is that given the large scope of
work, I need a release channel where people can try this work and provide
feedback before api is finalized. Kubernetes client lacks stabilization
branching strategy, which makes this contribution problematic. Finally the
way I structured the codebase makes kubernetes just one of underlying
implementations for what is essentially a 'live query' problem. In a sense
kubernetes integration is almost like a driver for EF. I'll update this pr
as I get closer to finishing it.

On Sat., Aug. 29, 2020, 6:09 p.m. Richard Simpson, <notifications@github.com>
wrote:

> Would it be possible to revive this PR in some fashion?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-client/csharp/pull/394#issuecomment-683347042>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINFWHUNYK2CUZ5XNYJON3SDF37ZANCNFSM4LYIPYWA>
> .
>

RichiCoder1 commented:
> This code will added as part of project SteelToe.io over the next few
months. Main reason is that I'm using newer apis which prevent me from
contributing them here. Second reason is that given the large scope of
work, I need a release channel where people can try this work and provide
feedback before api is finalized. Kubernetes client lacks stabilization
branching strategy, which makes this contribution problematic. Finally the
way I structured the codebase makes kubernetes just one of underlying
implementations for what is essentially a 'live query' problem. In a sense
kubernetes integration is almost like a driver for EF. I'll update this pr
as I get closer to finishing it.

Basically build it separately, stabilize, and then backport the k8s specific bits? Sounds good :+1:. If you could post here when there's SteelToe bits available, I'd be interested!

ddieruf commented:
So 10 months later, @macsux did you ever make any progress with the Steeltoe project?
